tests/integration/utils.py

```diff
@@ -162,86 +156,91 @@ def __get_reference_type(object_path, root_object_paths=None):
         elif 'type' in reference:
             return reference['type']
         elif '$ref' in reference:
+            nested_reference = reference['$ref']
             # Avoid infinite recursion
             if not root_object_paths:
-                root_object_paths.append(object_path)
-            if reference['$ref'] not in root_object_paths:
-                root_object_paths.append(reference['$ref'])
-                return __get_reference_type(reference['$ref'],
+                root_object_paths = [object_path]
+            if nested_reference not in root_object_paths:
+                root_object_paths.append(nested_reference)
+                return __get_reference_type(nested_reference,
                                             root_object_paths)
 
         return None
 
     # Helper function to check resource object schema
-    def __check_resource_schema(resource, null_value_allowed):
+    def __check_resource_schema(resource):
         # Check resource type
         self.assertEqual(resource['type'], schema['type']['enum'][0])
         # Check resource attributes
         actual_attributes = resource['attributes']
         expected_attributes = __get_schema_attributes()
-        __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed)
+        __check_attributes_schema(actual_attributes, expected_attributes)
 
     # Helper function to check error object schema
-    def __check_error_schema(error, null_value_allowed):
+    def __check_error_schema(error):
         # Check error attributes
         actual_attributes = error
         expected_attributes = schema
-        __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed)
+        __check_attributes_schema(actual_attributes, expected_attributes)
 
     # Helper function to check through all attributes
-    def __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed=False):
+    def __check_attributes_schema(actual_attributes, expected_attributes):
         for field, actual_value in actual_attributes.items():
             expected_attribute = expected_attributes[field]
             expected_type = __get_attribute_type(expected_attribute)
-            if not null_value_allowed or (actual_value and expected_type):
-                self.assertIsInstance(actual_value, expected_type)
+            if (actual_value and expected_type) or (
+              'nullable' in expected_attribute and
+               expected_attribute['nullable'] is False):
```

Comment:
What is `nullable` stands for here? I recalled we discuss this last week and I think the nullable keys should be configured via either the config file or the swagger file. I checked the openapi doc and found that [`nullable` attributes](https://swagger.io/docs/specification/data-models/data-types/#nullable) will only be supported after openapi 3.0, which means we cannot use it in our current implementation.

In addition, since lines 191-193 are multi-lines is-else statement, you can use `\` to break the line and improve the readability a bit. Also, for the condition `expected_attribute['nullable'] is False`, it can be changed to `not expected_attribute['nullable']`.